### DISCUSSION TITLE
Inserting images into the text flow

### DISCUSSION THREAD

**Discussion by paul-frost** [craftcms/ckeditor#243] [Ideas] [answered]
A very common use case is to have an image within the text flow to illustrate what is being referenced in the text. I'm thinking where there is a paragraph or more referencing something rather than a single line which could be handled by a caption.
<img width="250" alt="Example_of text_with_related-image" src="https://github.com/craftcms/ckeditor/assets/384378/fe2758e7-f241-4636-aaff-c0184ac8041f">

A common presentation would be the image taking up 50% of the width available and the text wrapping around it on larger screens. The css could control the exact layout/size based on available space.

The CKEditor website demo for image insertion has lots of options to choose the float direction and percentage of page width, why can't that be available in theCraft implementation?

The only option currently available is to create an 'entry type' for a CKEditor field that has multiple seperate fields:
Image - Asset field
Related text - CKEditor field
Position - Radio or Dropdown field
Width - Radio or Dropdown field
Caption - Plain Text field
Alt text - Plain Text field

But this also abstracts all that out into a seperate entry, and the related text is no longer part of the parent entry. I'm not sure if there could be any longer term issues with doing that but it feels wrong.

It also looks awful when you are editing the entry in the control panel as you don't have enough control as to how it is presented. If you have several entry type options included it looks very disjointed.

![CKEditor_content_builder](https://github.com/craftcms/ckeditor/assets/384378/780980fa-9391-4259-8751-ecbbe1ed3621)

Hopefully we can can come up with a nicer UI than this.

**Comment by i-just** [comment, marked as answer]
Hi, thanks for getting in touch. You can use various floating and wrapping options, but first, you need to add the necessary config to the “CKEditor Config” you’d like to use those in. By default, the “Insert image” toolbar item comes with the `toggleImageCaption` and `imageTextAlternative` options, but you can add various image styles and style options, too. Try adding the following to your selected CKEditor Config under “Config options”:
```
"image": {
    "toolbar": [
      "imageStyle:inline",
      "imageStyle:wrapText",
      "imageStyle:breakText",
      "|",
      "toggleImageCaption",
      "imageTextAlternative",
      "transformImage"
    ],
    "styles": {
      "options": [
        "inline",
        "alignLeft",
        "alignRight",
        "alignCenter",
        "alignBlockLeft",
        "alignBlockRight",
        "block",
        "side"
      ]
    }
  }
  ```
That’ll give you the ability to change the style and alignment of the inserted image:
<img width="1411" alt="Screenshot 2024-06-12 at 12 17 09" src="https://github.com/craftcms/ckeditor/assets/4500340/811de0a9-10f5-4b9f-9494-23726f6da5ea">


You can also add styles that will be used while editing content under “Custom Styles”. For example:
```
.image-style-align-right {
  max-width: 50%;
}
```

A complete list of options is available here: https://ckeditor.com/docs/ckeditor5/latest/api/module_image_imageconfig-ImageConfig.html#member-styles

You might also find this discussion helpful: https://github.com/craftcms/ckeditor/discussions/94

I hope this helps!

**Reply by paul-frost** [reply]
Thanks, I've just tried that, but there seems to be a problem. I choose the image and then select float right and it looks fine in the CKEditor field. Checking the source code it generated, it was:

```
<figure class="image image-style-align-right">
<img style="aspect-ratio:5616/3744;" src="https://trailquest5.ddev.site/local/images/massimo-sartirana-W_mBu-aC2m8-unsplash.jpg#asset:5:url" width="5616" height="3744">
</figure>
```

but after saving the entry and re-loading it changed to:

```
<figure class="image image-style-align-right">
<img style="aspect-ratio:1200/1200;" src="https://trailquest5.ddev.site/local/images/massimo-sartirana-W_mBu-aC2m8-unsplash.jpg#asset:5:url" alt="" width="1200" height="1200">
</figure>
```

**Reply by i-just** [reply]
That’s a feature of the HTML Purifier. There’s a [`HTML.MaxImgLength`](http://htmlpurifier.org/live/configdoc/plain.html#HTML.MaxImgLength) setting, and the default value is 1200.

You can change it by editing the json config file used by your field (HTML Purifier config files live under `config/htmlpurifier/`).

**Reply by paul-frost** [reply]
Thanks, that's fixed it, but it's a shame that it doesn't maintain the aspect ratio. But I normally limit the size up uploaded images in the first place using the [Image Resizer plugin](https://verbb.io/craft-plugins/image-resizer/docs/get-started/installation-setup), so it shouldn't be a problem. I was just using the CraftQuest training site to experiement and it had some very large images.
There are so many config options that aren't explained or obvious in the Craft docs.

**Reply by paul-frost** [reply]
Sorry to be a pain, but I've been trying to edit your code to only allow right align or block centered and although it saves the config OK, on the entry edit page I get a white arrow that doesn't open anything above the image instead of the icon menu choices. Can you  point me in the right direction?

**Reply by i-just** [reply]
Is this what you're after?
```
"image": {
    "styles": {
      "options": [
        "alignRight",
        "block"
      ]
    },
    "toolbar": [
      "imageStyle:alignRight",
      "imageStyle:block",
      "|",
      "toggleImageCaption",
      "linkImage",
      "transformImage"
    ]
  }
```
It also works without the `"styles"` part. You can find more info on how to configure images, including how to define your own styles and buttons, here: https://ckeditor.com/docs/ckeditor5/latest/features/images/images-styles.html#demo

**Reply by paul-frost** [reply]
Perfect, thanks.
I feel a lot more confident about using CKEditor for a new Craft 5 site. I was starting to despair at how bad the UX was for longform content compared to using a Matrix field in Craft 4.
